util/src/test/java/io/camunda/zeebe/util/startup/StartupProcessTest.java

```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.util.startup;
+
+import static io.camunda.zeebe.util.sched.future.TestActorFuture.completedFuture;
+import static io.camunda.zeebe.util.sched.future.TestActorFuture.failedFuture;
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.awaitility.Awaitility.await;
+import static org.mockito.Mockito.inOrder;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+import static org.mockito.Mockito.when;
+
+import io.camunda.zeebe.util.sched.ConcurrencyControl;
+import io.camunda.zeebe.util.sched.TestConcurrencyControl;
+import io.camunda.zeebe.util.sched.future.ActorFuture;
+import io.camunda.zeebe.util.sched.future.TestActorFuture;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CountDownLatch;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+@SuppressWarnings({"rawtypes", "unchecked"})
+class StartupProcessTest {
+  private static final Object STARTUP_CONTEXT =
+      new Object() {
+        @Override
+        public String toString() {
+          return "startupContext";
+        }
+      };
+
+  private static final Object SHUTDOWN_CONTEXT =
+      new Object() {
+        @Override
+        public String toString() {
+          return "shutdownContext";
+        }
+      };
+
+  private static final ConcurrencyControl TEST_CONCURRENCY_CONTROL = new TestConcurrencyControl();
+
+  static final class InvocationCountingStartupStep implements StartupStep<Object> {
+
+    private int startupInvocationCounter = 0;
+    private int shutdownInvocationCounter = 0;
+
+    int getStartupInvocationCounter() {
+      return startupInvocationCounter;
+    }
+
+    int getShutdownInvocationCounter() {
+      return shutdownInvocationCounter;
+    }
+
+    @Override
+    public String getName() {
+      return "InvocationCountingStartupStep";
+    }
+
+    @Override
+    public ActorFuture<Object> startup(final Object o) {
+      startupInvocationCounter++;
+      return completedFuture(o);
+    }
+
+    @Override
+    public ActorFuture<Object> shutdown(final Object o) {
+      shutdownInvocationCounter++;
+      return completedFuture(o);
+    }
+  }
+
+  static final class WaitingStartupStep implements StartupStep<Object> {
```

Comment:
:wrench: Generally I would opt for the lowest required visibility for any type, if only to discourage reuse unless someone has actually thought about whether it makes sense to reuse or not. Same for the other step implementations here. As for the `@Nested` ones, I'm not sure if junit can pick up private classes? But as they're just the tests, I would be very surprised someone would be tempted to reuse them.

```suggestion
  private static final class WaitingStartupStep implements StartupStep<Object> {
```

Reply:
Funny thing: When I make them private AND remove the static modifier, then Idea will also sort it to the end :smile:

Reply:
:smile: Problem solved haha